nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2ClientConnection.java

```diff
@@ -74,22 +118,93 @@ Http2ClientConnection connect() {
         }
 
         return this;
     }
 
+    private Queue<Http2FrameData> buffer(int streamId) {
+        return buffer.computeIfAbsent(streamId, i -> new ArrayDeque<>());
+    }
+
+    Http2FrameData readNextFrame(int streamId) {
+        try {
+            // Don't let streams to steal frame parts
+            // Allways read whole frame(frameHeader+data) at once
+            connectionLock.lock();
+            Http2FrameData polled = buffer(streamId).poll();
+            if (polled == null) {
+                return readNextFrameInternal(streamId);
+            }
+            return polled;
+        } finally {
+            connectionLock.unlock();
+        }
+    }
+
+    private Http2FrameData readNextFrameInternal(int streamId) {
+        this.reader.ensureAvailable();
+        BufferData frameHeaderBuffer = this.reader.readBuffer(FRAME_HEADER_LENGTH);
+        Http2FrameHeader frameHeader = Http2FrameHeader.create(frameHeaderBuffer);
+        frameHeader.type().checkLength(frameHeader.length());
+        BufferData data;
+        if (frameHeader.length() != 0) {
+            data = this.reader.readBuffer(frameHeader.length());
+        } else {
+            data = BufferData.empty();
+        }
+        if (streamId != frameHeader.streamId()) {
+            if (0 == frameHeader.streamId()) {
+                switch (frameHeader.type()) {
+                    case GO_AWAY -> {
+                        Http2GoAway http2GoAway = Http2GoAway.create(data);
+                        recvListener.frameHeader(helidonSocket, frameHeader);
+                        recvListener.frame(helidonSocket, http2GoAway);
+                        this.close();
+                        throw new IllegalStateException("Connection closed by the other side, error code: "
+                                + http2GoAway.errorCode()
+                                + " lastStreamId: " + http2GoAway.lastStreamId());
+                    }
+                    case SETTINGS -> {
+                        Http2Settings http2Settings = Http2Settings.create(data);
+                        recvListener.frameHeader(helidonSocket, frameHeader);
+                        recvListener.frame(helidonSocket, http2Settings);
+                        // §6.5.3 Settings Synchronization
+                        ackSettings();
+                        //TODO: settings
+                        return null;
+                    }
+                    case WINDOW_UPDATE -> {
+                        Http2WindowUpdate http2WindowUpdate = Http2WindowUpdate.create(data);
+                        recvListener.frameHeader(helidonSocket, frameHeader);
+                        recvListener.frame(helidonSocket, http2WindowUpdate);
+                        //TODO: flow-control
+                        return null;
+                    }
+                }
+            }
+
+            LOGGER.log(System.Logger.Level.DEBUG, "Deffering frame " + frameHeader.type() + " with streamId:"
```

Comment:
Random drive by comment: It seems to be wasteful to allocate a new String just to throw it away when DEBUG is not active. Is there a reason to not use the format String variant instead?

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.Logger.html#log(java.lang.System.Logger.Level,java.lang.String,java.lang.Object...)

Reply:
Still in progress, I am logging this for development purposes, logging will be top-notch before PR is ready for review :slightly_smiling_face:.